lib/Support/InstanceGraph.cpp

```diff
@@ -114,26 +116,24 @@ InstanceGraphNode *InstanceGraph::lookup(ModuleOpInterface op) {
   return lookup(cast<ModuleOpInterface>(op).getModuleNameAttr());
 }
 
-ModuleOpInterface
-InstanceGraph::getReferencedModuleImpl(InstanceOpInterface op) {
-  return lookup(op.getReferencedModuleNameAttr())->getModule();
-}
-
 void InstanceGraph::replaceInstance(InstanceOpInterface inst,
                                     InstanceOpInterface newInst) {
-  assert(inst.getReferencedModuleName() == newInst.getReferencedModuleName() &&
-         "Both instances must be targeting the same module");
-
-  // Find the instance record of this instance.
-  auto *node = lookup(inst.getReferencedModuleNameAttr());
-  auto it = llvm::find_if(node->uses(), [&](InstanceRecord *record) {
-    return record->getInstance() == inst;
-  });
-  assert(it != node->usesEnd() && "Instance of module not recorded in graph");
-
-  // We can just replace the instance op in the InstanceRecord without updating
-  // any instance lists.
-  (*it)->instance = newInst;
+  ArrayAttr instRefs = inst.getReferencedModuleNamesAttr();
+  assert(instRefs == newInst.getReferencedModuleNamesAttr() &&
+         "Both instances must be targeting the same modules");
+
+  // Replace all edges between the module of the instance and all targets.
+  for (Attribute targetNameAttr : inst.getReferencedModuleNamesAttr()) {
+    // Find the instance record of this instance.
+    auto *node = lookup(targetNameAttr.cast<StringAttr>());
+    auto it = llvm::find_if(node->uses(), [&](InstanceRecord *record) {
+      return record->getInstance() == inst;
```

Comment:
Can a module occur multiple times, should they all be updated (doesn't look like they collapsed to single edge, may have missed a representation constraint that this list is unique?)

Reply:
I suspect modules can appear multiple times with different config options.  e.g. {A -> foo, B -> foo, C -> bar}

Reply:
They can appear multiple times and an instance record is created for each occurrence. I adjusted the replace method to work with both options, but we haven't quite yet pinned down whether we want to de-duplicate here.